[13.0][MIG] connector_ecommerce: Migration to 13.0

victoralmau opened the pull request:
Superseed: https://github.com/OCA/connector-ecommerce/pull/59

Migration to 13.0

Locked by:
- [x] `sale_automatic_workflow_payment_mode` https://github.com/OCA/sale-workflow/pull/1673

Please @pedrobaeza and @joao-p-marques can you review it?

@Tecnativa TT30643

Commit:
[CHG] rename 'base_sale_multichannels' to 'connector_ecommerce'

[FIX] decimal_precision import does not use openerp.addons namespace

jgrandguillaume committed:
[ADD] Base support to export picking (partial and complete)

Commit:
[FIX] improve styling, remove unused imports, fix license (done by Joel)

[FIX] fix import

jgrandguillaume committed:
[ADD] Event to handle the adding of the tracking number on picking

Commit:
[FIX] remove the carrier tracking number from the event, fix missing import

jgrandguillaume committed:
[ADD] Support for exporting the account_invoice to Magento

sebastienbeau committed:
[IMP] start adding support of onchange in connector-ecommerce

[IMP] finish onchange support, add doc string, support tax. The option use external taxe will be not implemented, as re-computing tax in OpenERP is the best option. If you need it, create a independant module for it, fix copyright

[REF] refactor introduce the class SaleOrderOnChange

Commit:
[FIX] styling, license, imports, some fixes

[FIX] remove an old view not installable

bealdav committed:
[FIX] add product_view.xml file

Commit:
[REF] code review

sebastienbeau committed:
[IMP] code review

gurneyalex committed:
[IMP] play on change on sale.order and sale.order.lines in new framework, + support adding extra s.o. lines on the fly

add comments on the internals of SaleOrderOnChange.play()

Commit:
[FIX] wrong arguments in action_done

[IMP] work on stock.picking.out instead of stock.picking

[REM] clean delivery carrier old code

[ADD] a connector hook in the stock.picking.out view

[FIX] action_done is not called on stock.picking.out but only on stock.picking

[CHG] rename base_sale_data.xml to ecommerce_data.xml

[ADD] hook in invoice view for connectors

[ADD] account_invoice.origin_order_id, link to the order which created the invoice

[REF] moved sale.py code to old_stuff.py

[ADD] copy the order id to the invoice it creates

[REM] remove code

[REF] move content of partner.py to old_stuff.py

[REF] move partner_view.xml content to old_xml_stuff.xml

[REF] moved code from product.py to old_stuff.py

[REF] removed views and act_window links to ir.model.data from sale_view.xml

[REF] removed external.referential.category.csv

[REF] remove obsolete access groups

[REF] comment out the sale exception until its reimplementation

[REF] removed the import order wizard

[REF] removed connector.py

[FIX] style

[REF] moved external_referential.py code to old_stuff.py

[REF] remove empty files

[ADD] hook for connectors in sale view

[FIX] do not load removed file

[REF] updated __openerp__ manifest, changed copyright attribution to reflect the facts, changed the license from GPL to AGPL-3

[FIX] reintroduce methods used for the import of magento sale orders

[FIX] connector_ecommerce: replace dependencies on sale_quick_payment by sale_quick_payment

[ADD] connector_ecommerce: add import_rule and days_before_cancel in payment.method, theses fields will be used by the connectors

[IMP] connector_ecommerce: call the onchanges on the sale order to apply the workflow default values

[IMP] connector_ecommerce: activate the options 'Allow different addresses', 'Multiple shop', 'Discount on SO lines'

[FIX] noupdate=1 on the global settings activation

[FIX] missing name in a view

[IMP] add hints in payment method view to help in the choice of the import rules

[ADD] on_product_price_changed Event, fired when the price of a product could have changed

As the price is a fields.function based on the contextual pricelist, we can never be sure
that a price has changed, it depends on the context.

So we search all the base fields used to compute the price, and if one of these has changed,
the event is fired.

[FIX] initialize the context when None

[IMP] reviewed the connector_ecommerce's description in the manifest __openerp__.py

[FIX] remove obsolete sale exception

[ADD] fields cancelled_in_backend and cancellation_resolved to handle the cancellation of the sales orders from a backend

[IMP] add a notification when a user choose to not cancel a sales order cancelled on the backend

[ADD] notification when a sales order is cancelled on the backend

[IMP] ignore the cancellation from the backend should be done by a sales manager only

[IMP] ask for a reason when the sale order is not cancelled despite the cancellation on the backend

[FIX] filter on cancels should exclude the resolved cancels

[ADD] a parent_id field on the sale order, base implementation, the logic needs to be implemented in the connectors due to their own specificities

[ADD] button to open the parent sale order

[ADD] sale exception preventing confirmation of a sale order when it has a parent order cancelled on the backend

[IMP] add the sales manager rights to the connector manager because they usually need these rights

[IMP] fields 'need_cancel' and 'parent_need_cancel'

[FIX] spells cancelled with the US form: canceled

[FIX] typos in views

[FIX] hide the 'keep open' button in canceled state

[ADD] when we write 'canceled_in_backend' in a sales order, try to automatically cancel it

[FIX] filter out already cancelled sales orders from the filter 'to cancel'

[FIX] domain for the invisible attribute of the 'keep open' button, logic was reversed

[IMP] add a note on the related pickings and invoices of an order cancelled from backend

[FIX] cancelation of sales order: quotation states requires to be canceled via the workflow, sales order states with the action_cancel method. Some states can not be canceled.

[ADD] sale exception when an ordered product has a pending connector checkpoint

[ADD] french translation

Commit:
[ADD] 'Authorized' import rules to be able to import sales orders authorized by a payment institute but not paid yet.

Commit:
[REF] started to use ConnectorUnit classes to add extra lines in sale orders (shipping, ...), need a refactoring of the Mapper though. WIP

Commit:
Fix a bug : remove special fields before database storage (avoid to try to store non existing fields in db)

Commit:
[DEL] remove old stuff

mathieuvatel committed:
[IMP] - make possible to inherit the method _add_order_extra_line (done for prestashoperpconnect in a first place to be able to add or not the tax on delivery lines)

Commit:
[FIX] indentation according to pep8

Commit:
Added "invoice validated" event and called upon "on_invoice_validate".

Commit:
[FIX] the onchange test had a TypeError, now it fails for some reason

[ADD] a test which check if on_invoice_validated is correctly fired

[ADD] test the fire of 'on_invoice_paid'

[FIX] formatting of the onchange test

[FIX] fix the failing test: the payment_method_id and workflow_payment_id are not mandatory in a sale order

[FIX] fr_FR.po files not loaded, should be named fr.po

[FIX] context not propagated

[ADD] docstring

[CHG] rename ExtraOrderLineBuilder to SpecialOrderLineBuilder

[FIX] use session

[ADD] deprecation warning

[IMP] allow to use a product id or a product browse_record in lines builders

[FIX] Sales Orders onchanges: allow to have lines with other commands than (0, 0, {values}).

[FIX] simplify last change

[FIX] in lines builders, product_ref is not mandatory if we provide a product, renames 'product_id' to 'product', keeping ambiguity as we can have an int or browse_record.

[IMP] propagate the 'shop_id' in the context if it exists so the fiscal position rules (from openerp-fiscal-rules' project) can apply automatically

[FIX] 'int' object is not iterable error

[FIX] 'int' object is not iterable error, again

Commit:
Launchpad automatic translations update.

Commit:
[BUMP] release 2.1.0

Commit:
Launchpad automatic translations update.

Launchpad automatic translations update.

bealdav committed:
[IMP] move product_normal_form_view in connector_base_product

Commit:
Launchpad automatic translations update.

bealdav committed:
[FIX] product_view.xml in __openerp__.py

Commit:
Launchpad automatic translations update.

Commit:
The special sales orders lines can have a sequence and have low priority by default so they are likely to appear at the bottom, shipping line being the last one

rdeheele committed:
[FIX] related_backorder_ids shoud not be linked with a new copy

Commit:
rise the sequence of the special lines to reduce chances to be reached in case of orders with many lines, leave some space if people want to insert custom special lines in between

The default special's sequence is 980, for the same reason than before

do not build a new dict when default is an empty dict, not necessary to build a context dict (not used)

add read accesses on the sales users so they can confirm a sales order

bealdav committed:
[IMP] allow to override payment rules

Commit:
Bump to version 2.2.0

Commit:
Launchpad automatic translations update.

Launchpad automatic translations update.

Commit:
set modules to not installable

Move connector_ecommerce in the root folder and set installable

It now depends on sale_payment_automatic_workflow

In v7. sale_automatic_workflow had a dependency on sale_payment_method.
This is no longer true, now the module sale_payment_automatic_workflow
makes the glue between sale_automatic_workflow and sale_payment_method.

sale.shop has passed away

Remove the deprecated methods used to generate special lines, yeah!

stock.picking.out is no longer there

Remove the fiscal position rule onchange

As something similar has been implemented in v8, I'm not sure that
the fiscal position rule will be used and if it will, the onchange signature
will change, so let's remove it and add it again if necessary later.

Adapt the onchange players to the new API

Thanks to Model.onchange(), we apply the onchanges on a temporary 'new record' and get back the changed values.
Then we apply these values on our order values.

Use relative imports and remove fast_suite, checks in tests

Use the new env when building special lines

Rename environment to connector_env (follows connector's changes)

Use the new API

Correctly raise the event when a picking becomes 'done'

Test that on_tracking_number_added is fired

Correctly fire the event for partial pickings

Update the manifest and description; bump to 3.0.0

Correction of typos

Use @api.one as we have nothing outside the loop

Corrections of more typos

yvaucher committed:
Add bug tracker link on README.rst

Commit:
Trigger the change of price on the newly created product

kikopeiro committed:
[FIX] Bug throwing exception importing sale order

sbidoul committed:
[UPD] prefix versions with 8.0

pedrobaeza committed:
[MIG] Make modules uninstallable

parthiv committed:
[IMP] made module installable

[IMP] changes related to accounting views

[IMP] changes related to sales and stock views

[REN] renamed inherited class

[IMP] changes related to object rename

[FIX] indentation

[FIX] indentation with spaces

Commit:
Replace product.price.type by predefined fields

Remove workflow related code in sales cancelation

Rework onchange manager as we have only new API

remove +x on csv

Move line builder units in the unit directory

Use ConnectorSession.from_env() idiom

Fix OnChangeManager unit

* pass model to play_onchanges()
* extract id from (id, name) tuples on many2one fields
* get actual model and not 'model'

atchuthan committed:
[FIX] changed to MVC structure

[FIX] changed to MVC structure - modifications

[IMP] modifications for account_payment_mode

Commit:
Remove account.tax.group, now exists in 'account'

Improve phrasing, short license headers, ...

Rename class Environment to ConnectorEnvironment

Following the upstream change

atchuthan committed:
[FIX] renamed filename to account_payment_mode

[FIX] sale exception rule eval correction and xml odoo definition

Commit:
Capitalize products names for consistency

Correct tests of connector_ecommerce

Do not copy sale fields (+ _compute naming)

parthiv committed:
[FIX] removed duplicate line

[FIX] dependency on taxes_id for tax_group

[FIX] indentetion issue with python expression

[FIX] code optimization for tax

[REM] removed unnecessary attribute

[FIX] flake8 validation

Commit:
Extract onchange fields for better extensibility

Subclasses of the SaleOrderOnChange are now able to add or remove fields
from the list of fields on which the onchanges are applied.

pedrobaeza committed:
[MIG] Make modules uninstallable

[MIG] Rename manifest files

Commit:
[MIG] Migration to 10.0

sergio-teruel committed:
connector_ecommerce: Add security groups to avoid show information always.

lmignon committed:
[MIG] Migration to 10.0: Adapt code to 10.0

Commit:
fix message_post() expecting a len(recordset) == 1

Use new component system

Correct model name

Add new style events

Show connector tab only for connector manager

Fix indentation of views

Dispatch on_picking_out_done on stock.picking

Create sale shipping line as 'is_delivery'

And do not recreate it on delivery if it has been created in the first
place

Bump connector_ecommerce version

jaredkipe committed:
Fire new event: 'on_picking_dropship_done' when an 'incoming' shipment happens but is effectively a 'dropship'

Reduce duplicate code.

Simplify method expression.

Commit:
Fix lint error introduced in #43

hugosantosred committed:
[MIG] connector_ecommerce to V11

oca-travis committed:
[UPD] Update connector_ecommerce.pot

oca-transbot committed:
Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

ljimenezsidoo committed:
Connector Ecommerce to version 12

sebalix committed:
Update connector_ecommerce/views/sale_view.xml

Co-Authored-By: acm1pt-colorado <luisadanjimenez@gmail.com>

ljimenezsidoo committed:
Update connector_ecommerce/tests/test_picking_event.py

Co-Authored-By: Iván Todorovich <ivan.todorovich@druidoo.io>

ljimenezsidoo committed:
Update connector_ecommerce/tests/test_picking_event.py

Co-Authored-By: Iván Todorovich <ivan.todorovich@druidoo.io>

Commit:
RMV rule_group from ecommerce data xml

oca-travis committed:
[UPD] Update connector_ecommerce.pot

OCA-git-bot committed:
[ADD] icon.png

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: connector-ecommerce-12.0/connector-ecommerce-12.0-connector_ecommerce
Translate-URL: https://translation.odoo-community.org/projects/connector-ecommerce-12-0/connector-ecommerce-12-0-connector_ecommerce/

HviorForgeFlow committed:
[IMP] connector_ecommerce: black, isort

HviorForgeFlow committed:
[MIG] connector_ecommerce: Migration to 13.0

HviorForgeFlow committed:
[TMP] oca_dependencies.txt

bizzappdev committed:
[FIX]misc changes and onchange fix for sol

bizzappdev commented:
https://github.com/OCA/connector-ecommerce/pull/59 this is already there. why a new one?

victoralmau commented:
> #59 this is already there. why a new one?

The previous PR seemed to be abandoned because there was no response since 1 month ago and other changes needed to be added (only 1 migration commit, remove the custom branches in `oca_dependencies.txt` and add some changes that occurred when running `connector_magento` tests). In addition, the PR had been created without migrating its dependencies (`sale_automatic_workflow_payment_mode`).

pedrobaeza commented:
@bizzappdev do you agree then with the supersede?

joao-p-marques commented on connector_ecommerce/components/sale_order_onchange.py:
```diff
@@ -44,6 +45,9 @@ def play_onchanges(self, model, values, onchange_fields):
             )
             all_values.update(new_values)
 
+        for fld in not_values:
+            if fld not in all_values:
```
Will this condition ever be false? Before, you define both dictionaries based on the list of fields for a model, so AFAICS an element in one of them will never exist on the other, right?

That would make the check unnecessary, but I might be seeing it wrong… 😅

victoralmau replied:
I believe this has been clarified and is necessary due to `onchange_values`, right?

bizzappdev replied:
this is because after applying onchange. [HERE](https://github.com/OCA/connector-ecommerce/pull/62/files/80317cee37d322329c88c06e721cb152ccdd32e8..e9be4e72d5a27320328b1b3112bd2d4603011705#diff-655a37a2a57355d263c191d9d14a7f764edcb6b5a90a11d1a71eea8ef006317bR46) the all_values has been changed. and updated accordingly.

joao-p-marques reviewed (commented):
Code review. Overall looks good, just a small doubt.

victoralmau commented:
> Code review. Overall looks good, just a small doubt.

Thanks, please, can you review too `sale_automatic_workflow_payment_mode` ?

bizzappdev commented:
> supersede

I consider this as wast of resources and wast of time. and stealing the credits for contribution. 

why to reinvent wheel? things were straight forward in the another PR if something is broken then report there. 
We all as community should work as TEAM and TEAM Members. they can also create an MR for the Branch itself to my repo.
this MR literally has code from my commits.  I have also Superseed from the another MR which was inactive for more than 6 months. BUT I have kept their commits and history.

MY commits: 
1. https://github.com/bizzappdev/connector-ecommerce/commit/7f73f7c393c23f3d5b70ef5a0cc18e6df8799dbe
2. https://github.com/OCA/connector-ecommerce/pull/59/commits/00961baf758681c7a5b76bcedf2a3cead1782188#diff-655a37a2a57355d263c191d9d14a7f764edcb6b5a90a11d1a71eea8ef006317bR103

Their commit:
https://github.com/OCA/connector-ecommerce/pull/62/commits/e9be4e72d5a27320328b1b3112bd2d4603011705#diff-655a37a2a57355d263c191d9d14a7f764edcb6b5a90a11d1a71eea8ef006317bR98

I do not mind to contribute, but they are also no mention of any name, credits etc.
This kind of events and incidents will demotivate people to contribute.

pedrobaeza commented:
We don't want to steal anything, but your PR seems incomplete according my colleague. I see that my colleague @victoralmau has respected commit attribution from @HviorForgeFlow, that I suppose it was the most completed PR till the date. Maybe @victoralmau can add more information.

bizzappdev commented:
> We don't want to steal anything, but your PR seems incomplete according my colleague. I see that my colleague @victoralmau has respected commit attribution from @HviorForgeFlow, that I suppose it was the most completed PR till the date. Maybe @victoralmau can add more information.

yeah it was most complete MR and then I added the few changes for the Migration which was not done on previous one. 
for onchange. Please check the my commit and your colleague's commit. even the comments in that code block are also same.

pedrobaeza commented:
Thanks for the insights. @victoralmau please check the comments from @bizzappdev

bizzappdev commented:
> We don't want to steal anything, but your PR seems incomplete according my colleague. I see that my colleague @victoralmau has respected commit attribution from @HviorForgeFlow, that I suppose it was the most completed PR till the date. Maybe @victoralmau can add more information.

And in my initial comment I means "I consider" I meant by I feel like. I am not blaming. I am just point out.

pedrobaeza commented:
> And in my initial comment I means "I consider" I meant by I feel like. I am not blaming. I am just point out.

Yes, that's why we are trying to clarify things. What we want is to finish this with all the needed things and considerations.

bizzappdev commented:
> > And in my initial comment I means "I consider" I meant by I feel like. I am not blaming. I am just point out.
> 
> Yes, that's why we are trying to clarify things. What we want is to finish this with all the needed things and considerations.

yeah, but OCA has the policy for not removing commits or something related to commits for migration.
then why don't we have policy for this king of PR which were inspired or carry forwarded based some another PR.
it has to keep the original author's Commit and name

pedrobaeza commented:
Let me check with my colleague. Maybe it's just a mistake squashing things. If we wanted to act badly, we didn't ask you in first place.

victoralmau commented:
Sorry, in no case was intended to change anyone's contribution, just wanted to unify +10 commits in one migration commit as usual + add co-author. 
I have modified the PR to add all the commits that were in the mentioned PR and added a new one with some changes.
Is it correct now, right?

bizzappdev commented:
> Sorry, in no case was intended to change anyone's contribution, just wanted to unify +10 commits in one migration commit as usual + add co-author.
> I have modified the PR to add all the commits that were in the mentioned PR and added a new one with some changes.
> Is it correct now, right?

in that case I have squash my commit into one and you can rebase.

victoralmau committed:
[MIG] connector_ecommerce: Migration to 13.0

victoralmau commented:
Changes done.

victoralmau committed:
[IMP] connector_ecommerce: Add rule_paid_acquire_id so that if the import rule is defined as 'paid' in another addon that inherits from this one (connector_magento for example) we can use it to create a transaction for the paid amount of an order and the order will be auto-confirmed.

joao-p-marques commented on connector_ecommerce/components/sale_order_onchange.py:
```diff
@@ -93,8 +95,21 @@ def play(self, order, order_lines):
                     old_line_data = command_line[2]
+                    # passing order_id also so order_id.company_id based onchange will
+                    # also work
+                    old_line_data["order_id"] = new_record
```
Shouldn't this be like this?
```suggestion
                    old_line_data["order_id"] = new_record.id
```

joao-p-marques reviewed (commented):
@victoralmau just another doubt in @bizzappdev's change from code review

victoralmau committed:
[FIX] connector_ecommerce: Set order_id in play function according to sale_order_onchange component

victoralmau commented:
> @victoralmau just another doubt in @bizzappdev's change from code review

Thanks, changes done.

victoralmau commented:
Now `sale_automatic_workflow_payment_mode` addon is already merged.
I have relaunched Travis to verify that the tests work correctly.
If you confirm that everything is OK, it could already be merged, right?

pedrobaeza commented:
@bizzappdev is it OK for you?

bizzappdev commented:
> @bizzappdev is it OK for you?

Yeah LGTM

pedrobaeza commented:
Thanks. Let's merge it then

/ocabot merge nobump
